docs/rfc/rc-005-event-system.rst

```diff
@@ -0,0 +1,121 @@
+=====================
+RFC 005: Event System
+=====================
+
+Changelog
+---------
+
+- 2021-09-17: Initial Draft (@tychoish)
+
+Abstract
+--------
+
+The event system within Tendermint, which supports a lot of core
+functionality, also represents a major infrastructural liability. As part of
+our upcoming review of the RPC interfaces and our ongoing thoughts about
+stability and performance, as well as the preparation for Tendermint 1.0, we
+should revisit the design and implementation of the event system. This
+document discusses both the current state of the system 
+
+Background
+----------
+
+Current State of Events
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The event system makes it possible for clients, both internal and external,
+to receive notifications after a series of given events, such as new blocks,
+new transactions, validator set changes, as well as intermediate events during
+consensus. Because the event system is very cross cutting, the behavior and
+performance of the event publication and subscription system has huge impacts
+for all of Tendermint.
+
+The subscription service is exposed over the RPC interface, but also powers
+the indexing (e.g. to an external database,) and is the mechanism by which
+`BroadcastTxCommit` is able to wait for transactions to land in a block.
+
+The current pubsub mechanism relies on a couple of buffered channels,
+primarily between all event creators and subscribers, but also for each
+subscription. The result of this design is that, in some situations with the
+right collection of slow subscription consumers the event system can put
+backpressure on the consensus state machine and message gossiping in the
+network, thereby causing nodes to lag.
+
+Improvements
+~~~~~~~~~~~~
+
+The current system relies on implicit queues built by the buffered channels,
+and though threadsafe, can force all activity within Tendermint to serialize,
+which does not need to happen. Additionally, timeouts for subscription
```

Comment:
We do probably need to make some guarantees about order of events—if clients need the event stream to reflect the partial order of what's happened in consensus, we probably can't let the consensus state machine and the event stream get too far apart. So, for example, maybe we do have to stall if indexing is slow. But I very much agree that a slow subscriber shouldn't be able to stall consensus—and we should probably define some service level objectives for things like:

1. How fast you need to be able to consume messages.
2. What happens to messages if you don't go fast enough (drop messages? disconnect subscriber? maybe some combination depending on how slow things get?)
3. How much history you can expect to catch up on if you lose your connection to the stream, and how to cursor into the stream

Some of these things are kind of implementation defined right now.

Reply:
My (very rough) conjecture is that we don't actually need to publish very many events from consensus, and I think winnowing the types of events before we start implementing QoS features for the event system will be useful, but I think these are good ways to frame the requirements and objectives.